forge/test/mlir/llama/tests/test_llama_decode.py

```diff
@@ -0,0 +1,335 @@
+from transformers.modeling_attn_mask_utils import _prepare_4d_causal_attention_mask
+
+import forge
+from forge.op.eval.common import compare_with_golden_pcc
+
+
+class LlamaModelWrapper(torch.nn.Module):
+    """
+    In LlamaModelWrapper class, forward function takes input_ids (i.e list of input token or
+    or last predicted single token), attention_mask and return logits
+    Args:
+        1) use_cache = False
+            input_id (`torch.Tensor`) - shape of (batch_size, seq_length)
+            attention_mask (`torch.Tensor`) -  shape of (batch_size, seq_length)
+        2) use_cache = True
+            input_id (`torch.Tensor`) - shape of (batch_size, 1)
+            attention_mask (`torch.Tensor`) -  shape of (batch_size, key_value_seq_len + 1)
+            position_ids (`torch.Tensor`) -  shape of (batch_size, 1)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+                        key/values shape: (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+
+    Returns:
+        1) use_cache = False
+            logits (`torch.Tensor`) - shape (batch_size, seq_length, vocab_size)
+        2) use_cache = True
+            logits (`torch.Tensor`) - shape (batch_size, 1, vocab_size)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+                        key/values shape: (batch_size, num_of_key_values_heads, key_value_seq_len + 1, head_dim)
+    """
+
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.embed_tokens = model.model.embed_tokens
+
+    def forward(self, input_ids, attention_mask, position_ids=None, past_key_values=None):
+
+        inputs_embeds = self.embed_tokens(input_ids)
+
+        # The causal_mask calculated inside _update_causal_mask method in LlamaModel module
+        # is not properly traced by torch.jit.trace while converting pytorch to relay ops in TVM PyTorch Frontend Conversion
+        # which leads to pcc drops, because the causal mask helps to avoid attention calculation on padding token in LlamaAttention module
+        # To overcome this jit trace issue, causal mask is calculated by using _prepare_4d_causal_attention_mask function.
+        past_key_values_length = past_key_values[0][0].shape[-2] if past_key_values is not None else 0
+        causal_attention_mask = _prepare_4d_causal_attention_mask(
+            attention_mask, input_ids.shape, inputs_embeds, past_key_values_length
+        )
+        model_outputs = self.model(
+            attention_mask=causal_attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+        )
+
+        if model_outputs.past_key_values is not None:
+            return model_outputs.logits, model_outputs.past_key_values
+
+        return model_outputs.logits
+
+
+def load_model_and_tokenizer(model_path="openlm-research/open_llama_3b", **kwargs):
```

Comment:
Should this one be a common function "part of utilis.py" for Llama models?

Reply:
I have created the function inside the test_llama_decode.py purposely because the openlm-research/open_llama_3b variant will not generate correct token when we use autotokenizer, @mstojkovicTT mentioned fix https://github.com/tenstorrent/tt-forge-fe/issues/422#issuecomment-2441188184 here, but I can't find out the fix.  I don't want to distrub the existing code. so why I created function inside the test file.

Reply:
Can we merge the two to capture these details in one place? 

More precisely, having conditions per variant and applying different tokenizers is okay. Having those in a single place will make it much more easier to review and understand why there are these small differences

Reply:
Will update it

Reply:
Updated.